src/Database/PG/Query/Pool.hs

```diff
@@ -281,21 +286,48 @@ withExpiringPGconn pool f = do
-      when connectionStale $ do
-        -- Throwing is the only way to signal to resource pool to discard the
-        -- connection at this time, so we need to use it for control flow:
-        throw PGConnectionStale
-      -- else proceed with callback:
-      f connRsrc
-        -- FIXME this segfaults sometimes... see cbits/libpq-bindings.c
-        -- Clean up the connection buffers to prevent memory bloat (See #5087):
-        -- <* liftIO (unsafeClampInOutBufferSpace pgPQConn)
+    -- We would like to log the cases when the pool is completely saturated, This is 
+    -- not possible by only using 'RP.withResource' because when the pool is saturated it 
+    -- goes into the 'retry' loop and waits until atleast one resource is empty. Hence it's
+    -- not trivial to log the error from inside the STM 'retry' block.
+    -- So we use the 'tryWithResource' which returns 'Nothing' if connection is not possible,
+    -- when that happens we log the error and then use the 'RP.withResource' to establish the 
+    -- connection. This way we can log the error as well as execute the 'retry' logic.
+    resource <- RP.tryWithResource (_pool pool) $ \connRsrc -> action connRsrc old
+    case resource of
+      -- Log when the pool is saturated
+      Nothing -> do
+        liftIO $ _logger pool $ PLERetryMsg  "connection pool saturated... waiting to acquire connection"
+        RP.withResource (_pool pool) $ \connRsrc -> action connRsrc old  
+      (Just res) -> return res
+    where
+      action connRsrc@PGConn{..} old = do 
+        now <- liftIO getCurrentTime
+        let millis = realToFrac (1000000 * diffUTCTime now old)
+        liftIO (EKG.Distribution.add (_poolConnAcquireLatency (_stats pool)) millis)
+        let connectionStale =
+              maybe False (\lifetime-> now `diffUTCTime` pgCreatedAt > lifetime) pgMbLifetime
+
+        -- Log a warning when the pool is getting saturated
+        isPoolSaturated <- liftIO _isPoolSaturated 
+        when isPoolSaturated $ do
+          liftIO $ do pgLogger  $ PLEWarnMsg "connection pool nearing saturation, consider increasing connection pool size"
+        
+        when connectionStale $ do
+          -- Throwing is the only way to signal to resource pool to discard the
+          -- connection at this time, so we need to use it for control flow:
+          throw PGConnectionStale
+        -- else proceed with callback:
+        f connRsrc
+          -- FIXME this segfaults sometimes... see cbits/libpq-bindings.c
+          -- Clean up the connection buffers to prevent memory bloat (See #5087):
+          -- <* liftIO (unsafeClampInOutBufferSpace pgPQConn)
+      
+      _isPoolSaturated = do
+        inUseConns <- getInUseConnections pool
+        let maxConns =  RP.maxResources (_pool pool)
+        let saturationLevel = (fromIntegral inUseConns / fromIntegral maxConns) * 100.0 :: Double
+        return $ saturationLevel >= 80.0
 
```

Comment:
I feel that `withExpiringPGconn` is getting crowded, since it now interleaves four (mostly) separate pieces of functionality: pool saturation logging, pool saturation warnings, restarting "stale" connections, and logging connection acquisition times.

Maybe we could separate these pieces by structuring each as a function
```haskell
(PGPool -> (PGConn -> m a) -> m a) -> (PGPool -> (PGConn -> m a) -> m a)
```
as in the following (type-checking) snippet:
```haskell
withExpiringPGconn
  :: forall m a.
  (MonadBaseControl IO m, MonadIO m) => PGPool -> (PGConn -> m a) -> m a
withExpiringPGconn =
  -- Note: The behaviour of this composition depends on the order of the
  -- "filters", and I'm not sure what the "right" order is.
  destroyStaleConnectionFilter $
    acquireLatencyLoggingFilter $
      saturationWarningFilter $ \pool f -> do

  -- We would like to log the cases when the pool is completely saturated, This is 
  -- not possible by only using 'RP.withResource' because when the pool is saturated it 
  -- goes into the 'retry' loop and waits until atleast one resource is empty. Hence it's
  -- not trivial to log the error from inside the STM 'retry' block.
  -- So we use the 'tryWithResource' which returns 'Nothing' if connection is not possible,
  -- when that happens we log the error and then use the 'RP.withResource' to establish the 
  -- connection. This way we can log the error as well as execute the 'retry' logic.
  result <- RP.tryWithResource (_pool pool) f
  case result of
    -- Log when the pool is saturated
    Nothing -> do
      liftIO $ _logger pool $ PLERetryMsg  "connection pool saturated... waiting to acquire connection"
      RP.withResource (_pool pool) f
    Just res -> return res

-- Potential pitfall: Since such "filters" are so general (since we use MonadIO
-- m and MonadBaseControl IO m), their composition is not easy to reason about.
type WithConnFilter m a =
  (PGPool -> (PGConn -> m a) -> m a) ->
  (PGPool -> (PGConn -> m a) -> m a)

acquireLatencyLoggingFilter :: (MonadIO m) => WithConnFilter m a
acquireLatencyLoggingFilter withConn' pool f = do
  old <- liftIO getCurrentTime
  withConn' pool $ \conn -> do
    now <- liftIO getCurrentTime
    let microseconds = 1_000_000 * realToFrac (diffUTCTime now old)
    liftIO $ EKG.Distribution.add (_poolConnAcquireLatency (_stats pool)) microseconds
    f conn

saturationWarningFilter :: (MonadIO m) => WithConnFilter m a
saturationWarningFilter withConn' pool f =
  withConn' pool $ \conn -> do
    inUseConns <- liftIO $ getInUseConnections pool
    let maxConns = RP.maxResources (_pool pool)
        saturationLevel = (inUseConns * 100) `div` maxConns
        isNearingSaturation = saturationLevel >= 80
    when isNearingSaturation $
      liftIO $
        _logger pool $
          PLEWarnMsg "connection pool nearing saturation, consider increasing connection pool size"
    f conn

-- | Implementing a workaround for #5087, optionally expiring the connection
-- after a configurable amount of time so that memory at least can't accumulate
-- unbounded in long-lived connections.
--
-- See ticket for discussion of more long-term solutions.
--
-- Note that idle connections that aren't actively expired here will be
-- destroyed per the timeout policy in Data.Pool.
destroyStaleConnectionFilter ::
  (MonadBaseControl IO m, MonadIO m) => WithConnFilter m a
destroyStaleConnectionFilter withConn' pool f =
  -- If the connection was stale, we'll discard it and retry, possibly forcing
  -- creation of new connection:
  handleLifted
    (\PGConnectionStale -> destroyStaleConnectionFilter withConn' pool f) $
      withConn' pool $ \conn@PGConn{..} -> do

  now <- liftIO getCurrentTime
  let connectionStale =
        maybe False (\lifetime-> now `diffUTCTime` pgCreatedAt > lifetime) pgMbLifetime
  when connectionStale $ do
    -- Throwing is the only way to signal to resource pool to discard the
    -- connection at this time, so we need to use it for control flow:
    throw PGConnectionStale
  -- else proceed withConn' callback:
  f conn
    -- FIXME this segfaults sometimes... see cbits/libpq-bindings.c
    -- Clean up the connection buffers to prevent memory bloat (See #5087):
    -- <* liftIO (unsafeClampInOutBufferSpace pgPQConn)
```
Expressing `withExpiringPGconn` as the composition of such functions might be misleading to the reader, since their generality makes their composition hard to think about (especially when `MonadBaseControl IO m` is used to throw and catch exceptions). But the composition does at least separate namespaces, and would allow us to easily delete the part for restarting "stale" connections once that workaround becomes unnecessary.